### ISSUE TITLE
No Output

### ISSUE THREAD

**Issue by cocobln** [Aymkdn/tuyacloud-php#11]
Hello,

I hope you can help me. I try to find a solution for weeks :/

I m using Debian buster and there is no output, no matter which function I use.
There is no error in the error log and no information in the server logs.

I use the current Debian version and curl. All is up to date.

If I print_r the output of $tuya->getDevices() the Array is empty.

The general array for $tuya just shows my login data. The login data is correct.

 TuyaCloud Object
```php
 (
     [profile] => Array
         (
             [userName] => me@gmx.net
             [password] => XXXXXXXX
             [bizType] => smart_life
             [countryCode] => 49
             [region] => eu
             [from] => tuya
         )
 
     [uri] => https://px1.tuyaeu.com/homeassistant
     [accessToken] => XXXXXXXXXXXXXXXXXXXX
 )
```

Any idea what I can do that I can access my devices, linked with the tuya account and the tuya bridge?
All neccesary permissions are set in the tuya cloud.

Is there any debug mode for your script? Or is there any possibility to contact you somewhere? :)

Thank you very much :)

**Comment by Aymkdn** [comment]
There is no "debug mode", however you can add some `error_log` at different places to get some logs.

Because you get an accessToken (based on what you sent), then it means the login process is OK.

You could try to add a `error_log($dataStore->devices);` on https://github.com/Aymkdn/tuyacloud-php/blob/21f01c160f7401c074dbcda017a115759011ecdf/TuyaCloud.php#L117 and check what is returned by the Tuya server?

Are you sure you're using the "Smart Life" application on your phone, and not the Tuya application?

**Comment by cocobln** [comment]
There is nothing returned when I do that.

I found out, if I put an echo "Test" before line 50 it shows "test". If I put the echo after line 50, there is nothing. So I m not sure If the system really logs in. Maybe there is a problem within line 50.

And yes I use the "Smart Life" application :)

**Comment by Aymkdn** [comment]
Line 50 is a `return` https://github.com/Aymkdn/tuyacloud-php/blob/21f01c160f7401c074dbcda017a115759011ecdf/TuyaCloud.php#L50 so it's normal to have nothing after this line. 

You may have different line numbers if you put some additional code. But without sharing the line content, it's difficult to know what you're referring to.

**Comment by cocobln** [comment]
I put

error_log($dataStore->devices);

after

$dataStore->expire = time()+600; // every 600 seconds

---

There is no output. So I m not sure if the login really works. The output I mentioned in the beginning is generated even with wrong login data. Maybe its just a generated object before login. So I guess that there is no login. Otherwise there would be a difference with the wrong login data.

Any idea how to check if the login works?

Every try to output something in the TuyaCloud.php didn't work.

--

I tried 2 things:

  function login($retry=0) {
    // check if we have a valid token in memory
    $SHM_KEY = ftok(__FILE__, chr(1));
    $keyStore = crc32("tuyacloud_login");
    $store = shm_attach($SHM_KEY);

echo "something"; // outputs something

    if (shm_has_var($store, $keyStore)) {
      $dataStore = shm_get_var($store, $keyStore);
      $dataStore = json_decode($dataStore);
      if ($dataStore->expire > time()) {
        $this->accessToken = $dataStore->access_token;
        return;
      }
    }

echo "something"; // outputs nothing

    // if no data in memory, or token is not valid anymore
    $ch = curl_init($this->uri."/auth.do");

---

Thank you very much for your support.

**Comment by Aymkdn** [comment]
Do you call the file from q console or with a web server? From a server the `echo` might not be good, and `error_log` would write into the server's log.

Write something just after the `login($retry=0)` to see if it calls it at all. Because Tuya has a limit for the logins, once every minute. So there is a cache system. Maybe your system is not compatible with it?

**Comment by cocobln** [comment]
I call it from a web server in a php file. I used the example from the start page.

Writing something after login($retry=0) outputs something. Writing something before "// if no data in memory, or token is not valid anymore" outputs nothing.

**Comment by Aymkdn** [comment]
So it might be an issue with the cache system. Which version of php do you use?

**Comment by cocobln** [comment]
After

$this->accessToken = $dataStore->access_token;
return;

no output anymore in the whole login function. So I guess its something with it. So it doesn't come to start curl.

---

PHP Version is 7.3.31-1

**Comment by Aymkdn** [comment]
You can try to simplify the function by removing the part related to the cache. 

I'm sorry but I'm on my phone because it's already the night here, and tomorrow I'll go on vacation... And it's not easy to write code with a phone.

**Comment by cocobln** [comment]
I m not that good in removing that part. So maybe I should wait. I tried so many solutions and different APIs. I hoped that Tuya had a PHP API. But all the other stuff I don't understand :(

Maybe we can continue after the vacation.

Have a nice trip and good night :)

**Comment by Aymkdn** [comment]
Something like that? 

```php
function login($retry=0) {
    $ch = curl_init($this->uri."/auth.do");
    curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "POST");
    curl_setopt($ch, CURLOPT_POSTFIELDS, http_build_query($this->profile));
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch, CURLOPT_HTTPHEADER, array(
      'Content-Type: application/x-www-form-urlencoded'
    ));

    $data = curl_exec($ch);
    // $data is {"access_token":"xyz","refresh_token":"abc","token_type":"bearer","expires_in":864000}
    $tokens = json_decode($data);
    if (isset($tokens->access_token)) {
      // save access_token for 864000 seconds using shm_put_var() because we cannot login more than once in 60 seconds
      $dataStore = new stdClass();
      $dataStore->access_token = $tokens->access_token;
      $dataStore->expire = time()+$tokens->expires_in;
      $this->accessToken = $tokens->access_token;
    
      if ($retry === 5) throw new Exception("[tuyacloud] Unable to login: ".$data);
      else {
        sleep(1);
        return $this->login(++$retry);
    }
  }
```

**Comment by cocobln** [comment]
Thank you very much. I did the changes in Tuyacloud.php.

I made an output of the curl data. It shows that the curls seems to be called 2 times. The output is the following:

{"access_token":"EUheu163309XXXXXXXXXXXXXXXCqV4qhXN7","refresh_token":"EUheu16XXXXXXXXXXXXYuuemn3B7","token_type":"bearer","expires_in":864000}
{"responseStatus":"error","errorMsg":"you cannot auth exceed once in 60 seconds"}

So maybe it doesn't work because of 2 times in a line? I guess it is because the retry of 5 times.

**Comment by Aymkdn** [comment]
It's why I have this cache system, to avoid this error.

**Comment by cocobln** [comment]
Please go to bed and enjoy your vacation. If I don't find a solution I can try it after or I will send all the devices back to the seller :)

**Comment by Aymkdn** [comment]
Do you still need help?

**Comment by cocobln** [comment]
I set up a completely new vServer on another machine with Debian 9 stretch and php 7.3. An older version that we can exclude the cache error with the new version. Its the same situation. No output with correct login data. Even with wrong login data no output :/ I used the complete script and example from your website.

**Comment by Aymkdn** [comment]
> I made an output of the curl data. It shows that the curls seems to be called 2 times. The output is the following: 
> {"access_token":"EUheu163309XXXXXXXXXXXXXXXCqV4qhXN7","refresh_token":"EUheu16XXXXXXXXXXXXYuuemn3B7","token_type":"bearer","expires_in":864000} {"responseStatus":"error","errorMsg":"you cannot auth exceed once in 60 seconds"}
> 

This output is the second call. But what about the first call?

Try the below login function:
```php
function login($retry=0) {
    error_log("calling login with retry=".$retry);
    $ch = curl_init($this->uri."/auth.do");
    curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "POST");
    curl_setopt($ch, CURLOPT_POSTFIELDS, http_build_query($this->profile));
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch, CURLOPT_HTTPHEADER, array(
      'Content-Type: application/x-www-form-urlencoded'
    ));
    error_log("executing curl_exec...");
    $data = curl_exec($ch);
    error_log("result => ".$data);
    // $data is {"access_token":"xyz","refresh_token":"abc","token_type":"bearer","expires_in":864000}
    $tokens = json_decode($data);
    if (isset($tokens->access_token)) {
       error_log("token found: ".$tokens->access_token);
      // save access_token for 864000 seconds using shm_put_var() because we cannot login more than once in 60 seconds
      $dataStore = new stdClass();
      $dataStore->access_token = $tokens->access_token;
      $dataStore->expire = time()+$tokens->expires_in;
      $this->accessToken = $tokens->access_token;
    } else {
      error_log("No token found");
      if ($retry === 5) throw new Exception("[tuyacloud] Unable to login: ".$data);
      else {
        error_log("Trying to login again");
        sleep(1);
        return $this->login(++$retry);
      } 
    }
  }
```

**Comment by cocobln** [comment]
The first time no output and the second following:

Fatal error: Uncaught Exception: [tuyacloud] Unable to login: {"responseStatus":"error","errorMsg":"you cannot auth exceed once in 60 seconds"} in /var/www/html/tuyacloud-php/TuyaCloud.php:63 Stack trace: #0 /var/www/html/tuyacloud-php/TuyaCloud.php(67): TuyaCloud->login(5) #1 /var/www/html/tuyacloud-php/TuyaCloud.php(67): TuyaCloud->login(5) #2 /var/www/html/tuyacloud-php/TuyaCloud.php(67): TuyaCloud->login(4) #3 /var/www/html/tuyacloud-php/TuyaCloud.php(67): TuyaCloud->login(3) #4 /var/www/html/tuyacloud-php/TuyaCloud.php(67): TuyaCloud->login(2) #5 /var/www/html/tuyacloud-php/TuyaCloud.php(22): TuyaCloud->login() #6 /var/www/html/tuyacloud-php/index.php(10): TuyaCloud->__construct(Array) #7 {main} thrown in /var/www/html/tuyacloud-php/TuyaCloud.php on line 63

**Comment by Aymkdn** [comment]
This is not possible... If you replace the login function with what I sent you, you must have some logs...

**Comment by cocobln** [comment]
I replaced the login function completely by yours and this is the output. Thats why I wonder all the time.

Are there any requirements I need to install additionally that your script works? Is there any specific command I need to execute in the php example file? I just do the getDevices part.

If I replace it by this:
```php 
  function login($retry=0) {
    error_log("calling login with retry=".$retry);
    $ch = curl_init($this->uri."/auth.do");
    curl_setopt($ch, CURLOPT_CUSTOMREQUEST, "POST");
    curl_setopt($ch, CURLOPT_POSTFIELDS, http_build_query($this->profile));
    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
    curl_setopt($ch, CURLOPT_HTTPHEADER, array(
      'Content-Type: application/x-www-form-urlencoded'
    ));
    error_log("executing curl_exec...");
    $data = curl_exec($ch);
    error_log("result => ".$data);
    // $data is {"access_token":"xyz","refresh_token":"abc","token_type":"bearer","expires_in":864000}
    $tokens = json_decode($data);
    if (isset($tokens->access_token)) {
       error_log("token found: ".$tokens->access_token);
      // save access_token for 864000 seconds using shm_put_var() because we cannot login more than once in 60 seconds
      $dataStore = new stdClass();
      $dataStore->access_token = $tokens->access_token;
      $dataStore->expire = time()+$tokens->expires_in;
      $this->accessToken = $tokens->access_token;
    } else {
      error_log("No token found");
      if ($retry === 5) throw new Exception("[tuyacloud]â€¯Unable to login: ".$data);
      else {
        error_log("Trying to login again");
        sleep(1);
        return $this->login(++$retry);
    }
  }
```
---

I called it from the example file with:
```php 
require 'TuyaCloud.php';

$tuya = new TuyaCloud([
  "userName" => "mymail", // username/email to access to SmartLife/Tuya app
  "password" => "my password", // password to access to SmartLife/Tuya app
  "bizType" => "smart_life", // type ('tuya' or 'smart_life')
  "countryCode" => "33", // Country code (International dialing number), e.g. "33" for France or "1" for USA
  "region" => "eu" // region (az=Americas, ay=Asia, eu=Europe)
]);

// to get a list of your devices
$devices = $tuya->getDevices();

foreach($devices as $device) {
  echo "Name: ".$device->name."<br>";
  echo "ID: ".$device->id."<br>";
  echo "Type: ".$device->dev_type."<br>";
  if ($device->dev_type !== "scene") {
    echo "State: ".$device->data->state."<br>";
    echo "Online: ".$device->data->online."<br>";
  }
}
```
----

It shows:
> Parse error: syntax error, unexpected '?>', expecting function (T_FUNCTION) or const (T_CONST) in /var/www/html/tuyacloud-php/TuyaCloud.php on line 206

Then I put another } at the and of the login part and then it has the output I showed before.

The first execution has no output and if I reload I get:


> Fatal error: Uncaught Exception: [tuyacloud] Unable to login: {"responseStatus":"error","errorMsg":"you cannot auth exceed once in 60 seconds"} in /var/www/html/tuyacloud-php/TuyaCloud.php:63 Stack trace: #0 /var/www/html/tuyacloud-php/TuyaCloud.php(67): TuyaCloud->login(5) #1 /var/www/html/tuyacloud-php/TuyaCloud.php(67): TuyaCloud->login(5) #2 /var/www/html/tuyacloud-php/TuyaCloud.php(67): TuyaCloud->login(4) #3 /var/www/html/tuyacloud-php/TuyaCloud.php(67): TuyaCloud->login(3) #4 /var/www/html/tuyacloud-php/TuyaCloud.php(67): TuyaCloud->login(2) #5 /var/www/html/tuyacloud-php/TuyaCloud.php(22): TuyaCloud->login() #6 /var/www/html/tuyacloud-php/index.php(10): TuyaCloud->__construct(Array) #7 {main} thrown in /var/www/html/tuyacloud-php/TuyaCloud.php on line 63

**Comment by Aymkdn** [comment]
You don't need anything else. The php file is enough.

Just calling the below code is enough to run the login function :
```php
require 'TuyaCloud.php';

$tuya = new TuyaCloud([
  "userName" => "mymail", // username/email to access to SmartLife/Tuya app 
  "password" => "my password", // password to access to SmartLife/Tuya app 
  "bizType" => "smart_life", // type ('tuya' or 'smart_life')
 "countryCode" => "33", // Country code (International dialing number), e.g. "33" for France or "1" for USA 
 "region" => "eu" // region (az=Americas, ay=Asia, eu=Europe)
]);
```

Just doing the above will display plenty of things in your error log file with the login function I provided before (I added the missing `}`).
If it's not the case, then it means you're not including the correct PHP file...

**Comment by cocobln** [comment]
I thought you mean the output on the webpage. The following information I can see in the error log after the first execution:

[Tue Oct 12 19:38:23.142990 2021] [php7:notice] [pid 27909] [client ipaddress:50874] calling login with retry=0, referer: http://domainname/
[Tue Oct 12 19:38:23.143053 2021] [php7:notice] [pid 27909] [client ipaddress:50874] executing curl_exec..., referer: http://domainname/
[Tue Oct 12 19:38:23.242500 2021] [php7:notice] [pid 27909] [client ipaddress:50874] result => {"access_token":"EUheu16330examplexvGnNN","refresh_token":"EUheu16330exampleK6O1p8e5","token_type":"bearer","expires_in":864000}, referer: http://domainname/
[Tue Oct 12 19:38:23.242553 2021] [php7:notice] [pid 27909] [client ipaddress:50874] token found: EUheu1633exampleLVxhqC9xvGnNN, referer: http://domainname/

**Comment by Aymkdn** [comment]
Ok, so the login looks good. As you can see, we got the access token 👍

Next step is to change the `getDevices`:
```php
  function getDevices() {
    error_log("Calling getDevices");
    
    $body = [
      "header" => [
        "name" => 'Discovery',
        "namespace" => 'discovery',
        "payloadVersion" => 1
      ],
      "payload" => [
        "accessToken" => $this->accessToken
      ]
    ];
    $json = $this->post($this->uri.'/skill', $body);
    
    error_log("result : ".json_encode($json));

    return $json->payload->devices;
  }
```

And call it:
```php
require 'TuyaCloud.php';

$tuya = new TuyaCloud([
  "userName" => "mymail", // username/email to access to SmartLife/Tuya app
  "password" => "my password", // password to access to SmartLife/Tuya app
  "bizType" => "smart_life", // type ('tuya' or 'smart_life')
  "countryCode" => "33", // Country code (International dialing number), e.g. "33" for France or "1" for USA
  "region" => "eu" // region (az=Americas, ay=Asia, eu=Europe)
]);

// to get a list of your devices
$devices = $tuya->getDevices();

foreach($devices as $device) {
  echo "Name: ".$device->name."<br>";
  echo "ID: ".$device->id."<br>";
  echo "Type: ".$device->dev_type."<br>";
  if ($device->dev_type !== "scene") {
    echo "State: ".$device->data->state."<br>";
    echo "Online: ".$device->data->online."<br>";
  }
}
```

And check the error log.

**Comment by cocobln** [comment]
Thank you, I implemented it. This is the result:

[Tue Oct 12 21:05:12.210679 2021] [php7:notice] [pid 21876] [client ipaddress:52140] calling login with retry=0, referer: http://domainname/
[Tue Oct 12 21:05:12.210730 2021] [php7:notice] [pid 21876] [client ipaddress:52140] executing curl_exec..., referer: http://domainname/
[Tue Oct 12 21:05:12.293552 2021] [php7:notice] [pid 21876] [client ipaddress:52140] result => {"access_token":"EUheu16exampleANSEtWXr","refresh_token":"EUheu1examplevfSjCbzH","token_type":"bearer","expires_in":864000}, referer: http://domainname/
[Tue Oct 12 21:05:12.293601 2021] [php7:notice] [pid 21876] [client ipaddress:52140] token found: EUheu1exampleSEtWXr, referer: http://domainname/
[Tue Oct 12 21:05:12.294907 2021] [php7:notice] [pid 21876] [client ipaddress:52140] Calling getDevices, referer: http://domainname/
[Tue Oct 12 21:05:12.387498 2021] [php7:notice] [pid 21876] [client ipaddress:52140] result : {"payload":{"devices":[],"scenes":[]},"header":{"code":"SUCCESS","payloadVersion":1}}, referer: http://domainname/


---

It looks like there are no devices? I checked the tuya cloud and there are devices listed and linked with the account.

**Comment by Aymkdn** [comment]
I don't know... Did you try to replace the "bizType" from smart_life to tuya?

Other than that, I have no idea why it returns no device.

**Comment by cocobln** [comment]
The weird thing. The command line tool "tuya-cli" founds the devices. So it seems to be correct in my account.
But "tuya-cli" just works with the Cloud Application Authorization Key (Access ID/Client ID and Access Secret/Client Secret). Maybe Tuya changed something to their API.

I tried several things in the API Explorer of Tuya. The following function seems to work. 

This one:
![image](https://user-images.githubusercontent.com/946315/137082750-2e0a66ed-5c49-43ce-bb40-654dfc37be86.png)


Maybe you can include it to control a device by ID. I guess there was a change to the existing version in your php.

**Comment by Aymkdn** [comment]
My php file works perfectly for me. I get my devices without issue.

**Comment by Aymkdn** [comment]
I can see the different endpoints in https://github.com/TuyaAPI/openapi.

When I did this library I used the endpoints provided by home-assistant.

I have no plan to change my code because it's working very well for me. If it fails at some point, then I'll review to find a new solution.

Sorry but no change will be done.

**Comment by cocobln** [comment]
I guess there is another procedure to add the devices and permissions in the tuya cloud for your api.

do you think you could tell me the steps you did to register the devices at the tuya cloud for the usage with your script? Maybe thats the solution. 

Then i can delete the whole account at tuya and create a new one based on your steps. Maybe then it works with that procedure.

thank you very much :)

**Comment by Aymkdn** [comment]
I did nothing special: I open Smart Life application on my Android device and I add new device from there. I don't think I did anything else special!

**Comment by cocobln** [comment]
Tuya says an aplication needs be be created in their tuya cloud. Its very difficult. They changed many things. So maybe you have an old account and you didnt need to do anything in their cloud developer plattform. That would explain everything.

**Comment by Aymkdn** [comment]
Mmhhh I tried many things before making this library… So indeed, maybe I did something special.

It uses the API for Home Assistant, and on their page they ask for many different steps: https://www.home-assistant.io/integrations/tuya/

I probably tried these steps at some point… Maybe a lead for you?

**Comment by cocobln** [comment]
I deleted all of my tuya and smart_life accounts. After I used completely different email addresses to register the smart life account in the app and the tuya account for the cloud plattform. I exactly took the steps they mentioned on Homeassistent page you sent me. The login works, but there are no devices detected. I see all of my devices in the app and I see all of them on the tuya cloud platform. But not in the php request :/
